src/serac/numerics/functional/tests/tensor_unit_tests.cpp

```diff
@@ -152,3 +163,160 @@ TEST(tensor, implicit_conversion)
+      A[i][j].gradient = g[i][j];
+    }
+  }
+  auto [P, L, U] = lu(A);
+  auto PLU       = composeMatrixFromLU(P, L, U);
+
+  EXPECT_LT(squared_norm(get_value(A) - get_value(PLU)), tolerance);
+  EXPECT_LT(squared_norm(get_gradient(A) - get_gradient(PLU)), tolerance);
+}
+
+TEST(tensor, linear_solve_with_one_rhs)
+{
+  const tensor<double, 3, 3> A{{{2, 1, -1}, {-3, -1, 2}, {-2, 1, 2}}};
+  const tensor<double, 3>    b{{-1, 2, 3}};
+
+  auto x = linear_solve(A, b);
+  EXPECT_LT(squared_norm(dot(A, x) - b), tolerance);
+}
+
+TEST(tensor, linear_solve_with_multiple_rhs)
+{
+  const tensor<double, 3, 3> A{{{2, 1, -1}, {-3, -1, 2}, {-2, 1, 2}}};
+  const tensor<double, 3, 2> B{{{-1, 1}, {2, 1}, {3, -2}}};
+
+  auto X = linear_solve(A, B);
+  EXPECT_LT(squared_norm(dot(A, X) - B), tolerance);
+}
+
+TEST(tensor, derivative_of_linear_solve)
+{
+  // x defined by: t^2 * A * x(t) = t*b
+  // implicit derivative 2*t * A * x + t^2 * A * dxdt = b
+  // t^2 * A * dxdt = b - 2*t*A*x
+  // t^2 * A * dxdt = b - 2*t*b
+  // t = 1 --> dxdt = -x
+
+  const tensor<double, 3, 3> A{{{2, 1, -1}, {-3, -1, 2}, {-2, 1, 2}}};
+  const tensor<double, 3>    b{{-1, 2, 3}};
+
+  auto f = [&A, &b](dual<double> t) { return linear_solve(t * t * A, t * b); };
+
+  double t = 1.0;
+  auto   x = f(make_dual(t));
+
+  // expect x_dot = -x
+  EXPECT_LT(squared_norm(get_value(x) + get_gradient(x)), tolerance);
+}
+
+TEST(tensor, derivative_of_linear_solve_wrt_b_matches_finite_difference)
+{
+  const tensor<double, 3, 3> A{{{2, 1, -1}, {-3, -1, 2}, {-2, 1, 2}}};
+  tensor<double, 3>          b_value{{-1, 2, 3}};
+  const tensor<double, 3>    b_gradient{{0.337494265892494, 0.194238454581911, 0.307832573181341}};
+  auto                       b = make_dual(b_value, b_gradient);
+
+  auto x_value = linear_solve(A, b_value);
+
+  // first order forward difference
+  const double h     = 1e-6;
+  auto         dx_FD = (linear_solve(A, b_value + h * b_gradient) - x_value) / h;
+
+  auto x = linear_solve(A, b);
+
+  EXPECT_LT(squared_norm(dx_FD - get_gradient(x)), tolerance);
+}
+
+TEST(tensor, derivative_of_linear_solve_wrt_A_matches_finite_difference)
+{
+  const tensor<double, 3, 3> v{{{2, 1, -1}, {-3, -1, 2}, {-2, 1, 2}}};
+  const tensor<double, 3, 3> g{{{0.337494265892494, 0.194238454581911, 0.307832573181341},
+                                {0.090147365480304, 0.610402517912401, 0.458978918716148},
+                                {0.689309323130592, 0.198321409053159, 0.901973313462065}}};
+  tensor<dual<double>, 3, 3> A{};
+  for (int i = 0; i < 3; i++) {
+    for (int j = 0; j < 3; j++) {
+      A[i][j].value    = v[i][j];
+      A[i][j].gradient = g[i][j];
+    }
+  }
+  const tensor<double, 3> b{{-1, 2, 3}};
+
+  // central difference (2nd order accurate)
+  const double h     = 1e-6;
+  auto         dx_FD = (linear_solve(v + h * g, b) - linear_solve(v - h * g, b)) / (2 * h);
+
+  auto x = linear_solve(A, b);
+
+  EXPECT_LT(squared_norm(dx_FD - get_gradient(x)), tolerance);
+}
```

Comment:
one other test you might add is one where you try the LU and linear solve routines in a `constexpr` context, something like

```cpp
constexpr tensor<double, 3, 3> A{{{2, 1, -1}, {-3, -1, 2}, {-2, 1, 2}}};
constexpr tensor<double, 3>    b{{-1, 2, 3}};
constexpr auto x = linear_solve(A, b);
constexpr auto PLU = lu(A);
```

This will help verify that the constexpr evaluation works as intended too. Another benefit is that since undefined behavior (e.g. out-of-bounds array accesses, reading from uninitialized memory, etc.) is forbidden in a `constexpr` context, the compiler will locate any problems like that and let you know the exact line of code where it occurs.

Reply:
Good idea. I added it (see line 259). This led me to add missing `constexpr` specifiers to a couple functions in `tensor.hpp` and `dual.hpp`.